crypto/fipsmodule/bn/montgomery.c

```diff
@@ -119,8 +119,47 @@
 #include <openssl/type_check.h>
 
 #include "internal.h"
+#include "../cpucap/internal.h"
 #include "../../internal.h"
 
+#if !defined(OPENSSL_NO_ASM) &&                                                \
+    (defined(OPENSSL_LINUX) || defined(OPENSSL_APPLE)) &&                      \
+    defined(OPENSSL_AARCH64) && !defined(MY_ASSEMBLER_IS_TOO_OLD_FOR_AVX) &&   \
+    defined(OPENSSL_BN_ASM_MONT)
+
+#include "../../../third_party/s2n-bignum/include/s2n-bignum_aws-lc.h"
+
+#define BN_MONTGOMERY_USE_S2N_BIGNUM 1
+
+#endif
+
+OPENSSL_INLINE int montgomery_s2n_bignum_capable(void) {
+#if defined(BN_MONTGOMERY_USE_S2N_BIGNUM)
+
+  return 1;
+
+#else
+
+  return 0;
+
+#endif
+}
+
+OPENSSL_INLINE int montgomery_use_s2n_bignum(unsigned int num) {
+#if defined(BN_MONTGOMERY_USE_S2N_BIGNUM)
+
+  // Use s2n-bignum's functions only if (1) the ARM architecture has slow
+  // multipliers, and (2) temporary buffer's size does not exceed
```

Comment:
I see two more conditions in the code, can we document them as well? And also explain the `(2 * (uint64_t)num + 96) <= BN_MONTGOMERY_MAX_WORDS` part -- why `2*` and why `96`?

Reply:
The buffer size (2*num+96) is the arrays' size allocated in `montgomery_s2n_bignum_mul_mont()`, and `num % 8 == 0` is the precondition that s2n-bignum's extended montgomery reduction should work.
I will address these comments in the following commits.